File: src/main/java/com/example/demo/service/CalendarHolidayService.java
package com.example.demo.service;

import com.example.demo.mapper.CalendarHolidayMapper;
import com.example.demo.model.CalendarHoliday;
import com.example.demo.util.DbUtils;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class CalendarHolidayService {

    private final CalendarHolidayMapper calendarHolidayMapper;

    public CalendarHolidayService(CalendarHolidayMapper calendarHolidayMapper) {
        this.calendarHolidayMapper = calendarHolidayMapper;
    }

    @Cacheable(value = "holidays", key = "#year")
    public List<CalendarHoliday> getHolidaysByYear(int year) {
        return DbUtils.dbCall("selectByYear", () -> calendarHolidayMapper.selectByYear(year), year);
    }

    public List<CalendarHoliday> getHolidaysByRange(LocalDate from, LocalDate to) {
        if (from.isAfter(to)) {
            throw new IllegalArgumentException("from は to より後ろにできません");
        }
        return DbUtils.dbCall("selectByDateRange", () -> calendarHolidayMapper.selectByDateRange(from, to), from, to);
    }

    @Cacheable(value = "holidaysMap", key = "#year")
    public Map<String, String> getHolidaysMapByYear(int year) {
        List<CalendarHoliday> holidays = getHolidaysByYear(year);
        return holidays.stream().collect(Collectors.toMap(
                h -> h.getHolidayDate().toString(),
                CalendarHoliday::getName
        ));
    }

    public boolean isHoliday(LocalDate date) {
        return DbUtils.dbCall("selectByDate", () -> calendarHolidayMapper.selectByDate(date), date) != null;
    }

    @Transactional
    @CacheEvict(value = {"holidays", "holidaysMap"}, allEntries = true)
    public CalendarHoliday addHoliday(LocalDate date, String name) {
        CalendarHoliday holiday = new CalendarHoliday();
        holiday.setHolidayDate(date);
        holiday.setName(name);
        holiday.setYear(date.getYear());
        DbUtils.dbCall("insert", () -> calendarHolidayMapper.insert(holiday), date, name);
        return holiday;
    }

    @Transactional
    @CacheEvict(value = {"holidays", "holidaysMap"}, allEntries = true)
    public int updateHoliday(Long id, String name) {
        CalendarHoliday holiday = new CalendarHoliday();
        holiday.setId(id);
        holiday.setName(name);
        return DbUtils.dbCall("update", () -> calendarHolidayMapper.update(holiday), id, name);
    }

    @Transactional
    @CacheEvict(value = {"holidays", "holidaysMap"}, allEntries = true)
    public int deleteHoliday(Long id) {
        return DbUtils.dbCall("deleteById", () -> calendarHolidayMapper.deleteById(id), id);
    }
}

File: src/main/kotlin/com/example/demo/service/CalendarHolidayService.kt
package com.example.demo.service

import com.example.demo.mapper.CalendarHolidayMapper
import com.example.demo.model.CalendarHoliday
import com.example.demo.util.dbCall
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

@Service
class CalendarHolidayService(
    private val calendarHolidayMapper: CalendarHolidayMapper
) {
    /**
     * 指定した年の祝日一覧を取得（キャッシュ有効）
     */
    @Cacheable("holidays", key = "#year")
    fun getHolidaysByYear(year: Int): List<CalendarHoliday> {
        return dbCall("selectByYear", year) { calendarHolidayMapper.selectByYear(year) }
    }

    /**
     * 日付範囲で祝日を取得
     */
    fun getHolidaysByRange(from: LocalDate, to: LocalDate): List<CalendarHoliday> {
        require(!from.isAfter(to)) { "from は to より後ろにできません" }
        return dbCall("selectByDateRange", from, to) { calendarHolidayMapper.selectByDateRange(from, to) }
    }

    /**
     * 年指定で祝日をMap形式で取得（フロントエンド用、キャッシュ有効）
     * キー: 日付文字列(YYYY-MM-DD)、値: 祝日名
     */
    @Cacheable("holidaysMap", key = "#year")
    fun getHolidaysMapByYear(year: Int): Map<String, String> {
        val holidays = getHolidaysByYear(year)
        return holidays.associate { it.holidayDate.toString() to it.name }
    }

    /**
     * 特定の日付が祝日かどうか確認
     */
    fun isHoliday(date: LocalDate): Boolean {
        return dbCall("selectByDate", date) { calendarHolidayMapper.selectByDate(date) } != null
    }

    /**
     * 祝日を追加（キャッシュをクリア）
     */
    @Transactional
    @CacheEvict(value = ["holidays", "holidaysMap"], allEntries = true)
    fun addHoliday(date: LocalDate, name: String): CalendarHoliday {
        val holiday = CalendarHoliday(
            holidayDate = date,
            name = name,
            year = date.year
        )
        dbCall("insert", date, name) { calendarHolidayMapper.insert(holiday) }
        return holiday
    }

    /**
     * 祝日を更新（キャッシュをクリア）
     */
    @Transactional
    @CacheEvict(value = ["holidays", "holidaysMap"], allEntries = true)
    fun updateHoliday(id: Long, name: String): Int {
        return dbCall("update", id, name) {
            calendarHolidayMapper.update(CalendarHoliday(id = id, holidayDate = LocalDate.now(), name = name, year = 0))
        }
    }

    /**
     * 祝日を削除（キャッシュをクリア）
     */
    @Transactional
    @CacheEvict(value = ["holidays", "holidaysMap"], allEntries = true)
    fun deleteHoliday(id: Long): Int {
        return dbCall("deleteById", id) { calendarHolidayMapper.deleteById(id) }
    }
}

